lib/virtual-net/src/ruleset.rs

```diff
@@ -0,0 +1,1388 @@
+            (RuleType::Dns, rem)
+        } else if let Some(rem) = input.strip_prefix("ipv4:") {
+            (RuleType::IPV4, rem)
+        } else if let Some(rem) = input.strip_prefix("ipv6:") {
+            (RuleType::IPV6, rem)
+        } else {
+            return Err(RulesetError::InvalidRuleType(input.to_string()));
+        };
+
+        Ok(pair)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum RuleAction {
+    Allow,
+    Deny,
+}
+
+impl RuleAction {
+    pub fn consume_input(input: &str) -> Result<(Self, &str), RulesetError> {
+        let pair = if let Some(rem) = input.strip_prefix("allow=") {
+            (RuleAction::Allow, rem)
+        } else if let Some(rem) = input.strip_prefix("deny=") {
+            (RuleAction::Deny, rem)
+        } else {
+            return Err(RulesetError::InvalidRuleAction(input.to_string()));
+        };
+
+        Ok(pair)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct RuleExpr(String);
+
+impl RuleExpr {
+    pub fn consume_input(input: &str) -> Result<(Self, &str), RulesetError> {
+        let mut next_dns_entry = usize::MAX;
+        let mut next_ipv4_entry = usize::MAX;
+        let mut next_ipv6_entry = usize::MAX;
+
+        if let Some(idx) = input.find(",dns:") {
+            next_dns_entry = idx;
+        }
+
+        if let Some(idx) = input.find(",ipv4:") {
+            next_ipv4_entry = idx;
+        }
+
+        if let Some(idx) = input.find(",ipv6:") {
+            next_ipv6_entry = idx;
+        }
+
+        let next_entry = next_dns_entry
+            .min(next_ipv4_entry)
+            .min(next_ipv6_entry)
+            .min(input.len());
+
+        let (expr, rem) = input.split_at(next_entry);
+
+        let rem = rem.strip_prefix(',').unwrap_or(rem);
+
+        Ok((RuleExpr(expr.to_string()), rem))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct RulesetSegment {
+    ty: RuleType,
+    action: RuleAction,
+    expr: RuleExpr,
+}
+
+fn ruleset_segments(s: impl AsRef<str>) -> Result<Vec<RulesetSegment>, RulesetError> {
+    let mut input = s.as_ref();
+    let mut segments = Vec::new();
+
+    while !input.is_empty() {
+        let (ty, remaining) = RuleType::consume_input(input)?;
+        let (action, remaining) = RuleAction::consume_input(remaining)?;
+        let (expr, remaining) = RuleExpr::consume_input(remaining)?;
+
+        segments.push(RulesetSegment { ty, action, expr });
+
+        input = remaining;
+    }
+
+    Ok(segments)
+}
+
+#[derive(Debug, Clone)]
+pub struct Ruleset {
+    rules: Arc<RwLock<Vec<Rule>>>,
+}
+
+impl Ruleset {
+    pub fn is_socket_allowed(&self, addr: impl Into<SocketAddr>, dir: Direction) -> bool {
+        let addr = addr.into();
+
+        let ruleset = self.rules.read().unwrap();
+        let is_whitelisted = ruleset.iter().any(|r| r.is_socket_allowed(addr, dir));
+        let is_blacklisted = ruleset.iter().any(|r| r.is_socket_blocked(addr, dir));
+        drop(ruleset);
```

Comment:
nit: this isn't really needed, since all that follows is a single bool comparison.

But, for future reference, it's super easy to lose the `drop` during refactors, or do changes that extend the lock lifetime.

What I like to do is a nested scope:

```rust
let x = {
  // Nested scope to limit lock lifetime.
  let lock = self.y.lock().unwrap();
  ...
};
```